File: ui/src/main/java/com/peakmain/ui/compress/ImageCompressUtils.java
package com.peakmain.ui.compress;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.support.annotation.UiThread;
import android.text.TextUtils;
import android.util.Log;

import com.peakmain.ui.BuildConfig;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * author ：Peakmain
 * createTime：2020/3/28
 * mail:2726449200@qq.com
 * describe：图片压缩工具
 */
public class ImageCompressUtils implements Handler.Callback {
    //开始
    private static final int COMPRESS_START = 101;
    //成功
    private static final int COMPRESS_SUCCESS = 102;
    //失败
    private static final int COMPRESS_ERROR = 103;

    private static final String DEFAULT_DISK_CACHE_DIR = "basic_disk_cache";
    //所有图片的路径
    private List<String> mPaths;
    private int mLeastCompressSize;
    private String mOutFileDir;
    private OnCompressListener mCompressListener;
    private Handler mHandler;
    private int mQuality;

    public ImageCompressUtils(Builder builder) {
        this.mPaths = builder.mPaths;
        this.mLeastCompressSize = builder.mLeastCompressSize;
        this.mOutFileDir = builder.mOutFileDir;
        this.mCompressListener = builder.mCompressListener;
        mQuality = builder.mQuality;
        mHandler = new Handler(Looper.getMainLooper(),this);
    }

    public static Builder with(Context context) {
        return new Builder(context);
    }

    private File getImageCacheDir(Context context) {
        return getImageCacheDir(context, DEFAULT_DISK_CACHE_DIR);
    }

    private File getImageCacheDir(Context context, String cacheName) {
        File cacheDir = context.getExternalCacheDir();
        if (cacheDir != null) {
            File result = new File(cacheDir, cacheName);
            if (!result.mkdirs() && (!result.exists() || !result.isDirectory())) {
                return null;
            }
            return result;
        }
        Log.e(BuildConfig.TAG, "default disk cache dir is null");
        return null;
    }

    private File getImageCacheFile(Context context, String suffix) {
        if (TextUtils.isEmpty(mOutFileDir)) {
            mOutFileDir = getImageCacheDir(context).getAbsolutePath();
        }

        String cacheBuilder = mOutFileDir + "/" +
                System.currentTimeMillis() +
                (int) (Math.random() * 1000) +
                (TextUtils.isEmpty(suffix) ? ".jpg" : suffix);

        return new File(cacheBuilder);
    }

    @UiThread
    private void startCompress(final Context context) {
        if (mPaths == null || mPaths.size() == 0 && mCompressListener != null) {
            mCompressListener.onError(new NullPointerException("image file cannot be null"));
        }
        //存储压缩后的图片集合
        final List<String> result = new ArrayList<>();
        //原始图片的集合
        List<String> originalPath = new ArrayList<>(mPaths);
        mHandler.sendMessage(mHandler.obtainMessage(COMPRESS_START));

        for (final String path : originalPath) {
            if (Checker.isImage(path)) {
                //如果是图片
                //保证 同一时间只有一个Task在后台运行
                AsyncTask.SERIAL_EXECUTOR.execute(new Runnable() {
                    @Override
                    public void run() {
                        boolean isNeedCompress = Checker.isNeedCompress(mLeastCompressSize, path);
                        File file;
                        if (isNeedCompress) {
                            CompressUtils compressUtils = new CompressUtils();
                            Bitmap bitmap = compressUtils.decodeFile(path);
                            /* file = getImageCacheFile(context, Checker.checkSuffix(path));*/
                            compressUtils
                                    .compressImage(bitmap, mQuality, path);

                        } else {
                            file = new File(path);
                        }
                        if (!result.contains(path)) {
                            //添加到已经缓存的集合中
                            result.add(path);
                        }
                        //判断压缩是否已经结束
                        if (mPaths.size() == result.size()) {
                            mHandler.sendMessage(mHandler.obtainMessage(COMPRESS_SUCCESS, result));
                        }
                    }
                });
            } else {
                mCompressListener.onError(new IllegalArgumentException("This path is not a picture and cannot be read..."));
                break;
            }
        }
    }

    @Override
    public boolean handleMessage(Message msg) {
        if (mCompressListener == null) {
            return false;
        }
        switch (msg.what) {
            case COMPRESS_START:
                mCompressListener.onStart();
                break;
            case COMPRESS_ERROR:
                mCompressListener.onError((Throwable) msg.obj);
                break;
            case COMPRESS_SUCCESS:
                Log.e(BuildConfig.TAG, "成功了");
                mCompressListener.onSuccess((List<String>) msg.obj);
                break;
            default:
                break;
        }
        return false;
    }

    public static class Builder {
        private Context mContext;
        private List<String> mPaths;
        private int mLeastCompressSize = 100;
        private OnCompressListener mCompressListener;
        //输出文件目录
        private String mOutFileDir;
        //默认100
        private int mQuality = 100;

        public Builder(Context context) {
            this.mContext = context;
            this.mPaths = new ArrayList<>();
        }

        public Builder load(String path) {
            this.mPaths.add(path);
            return this;
        }

        public Builder setQuality(int quality) {
            mQuality = quality;
            return this;
        }

        public Builder load(List<String> paths) {
            this.mPaths.addAll(paths);
            return this;
        }

        public Builder setCompressListener(OnCompressListener listener) {
            this.mCompressListener = listener;
            return this;
        }

        public Builder setOutFileDir(String outFileDir) {
            this.mOutFileDir = outFileDir;
            return this;
        }

        //小于100，忽略图片压缩
        public Builder ignoreCompress(int size) {
            this.mLeastCompressSize = size;
            return this;
        }

        private ImageCompressUtils build() {
            return new ImageCompressUtils(this);
        }

        public void startCompress() {
            build().startCompress(mContext);
        }

    }


}

File: ui/src/main/java/com/peakmain/ui/compress/ImageCompressUtils.kt
package com.peakmain.ui.compress

import android.content.Context
import android.os.AsyncTask
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.TextUtils
import androidx.annotation.UiThread
import com.peakmain.ui.utils.FileUtils
import com.peakmain.ui.utils.FileUtils.getFileByPath
import com.peakmain.ui.utils.FileUtils.writeFileFromIS
import com.peakmain.ui.utils.LogUtils
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException

/**
 * author ：Peakmain
 * createTime：2020/3/28
 * mail:2726449200@qq.com
 * describe：图片压缩工具
 */
class ImageCompressUtils(builder: Builder) : Handler.Callback {
    //所有图片的路径
    private var mPaths: List<String>?
    private val mLeastCompressSize: Int
    private val mOutFileDir: String?
    private val mCompressListener: OnCompressListener?
    private val mHandler: Handler
    private val mQuality: Int

    @UiThread
    private fun startCompress() {
        if (mPaths == null || mPaths?.size == 0 && mCompressListener != null) {
            mCompressListener!!.onError(NullPointerException("image file cannot be null"))
        }
        //存储压缩后的图片集合
        val result: MutableList<String> = ArrayList()
        //原始图片的集合
        val originalPath: List<String> = ArrayList(mPaths)
        mHandler.sendMessage(mHandler.obtainMessage(COMPRESS_START))
        if (TextUtils.isEmpty(mOutFileDir)) {
            compressFile(result, originalPath)
        } else {
           if( FileUtils.createOrExistsDir(mOutFileDir)){
               LogUtils.d("文件目录存在或创建成功$mOutFileDir")
           }
            val compressFiles: MutableList<String> = ArrayList()
            for (path in mPaths!!) {
                val file = getFileByPath(path)
                val destFile = File(mOutFileDir + File.separator + file!!.name)
                if (file.isFile) {
                    //拷贝到指定的目录
                    try {
                        writeFileFromIS(destFile, FileInputStream(file))
                        compressFiles.add(destFile.absolutePath)
                    } catch (e: FileNotFoundException) {
                        e.printStackTrace()
                    }
                }
            }
            //进行的是拷贝之后路径的图片压缩
            compressFile(result, compressFiles)
        }
    }

    private fun compressFile(result: MutableList<String>, paths: List<String>) {
        for (path in paths) {
            if (Checker.isImage(path)) {
                //如果是图片
                //保证 同一时间只有一个Task在后台运行
                AsyncTask.SERIAL_EXECUTOR.execute {
                    val isNeedCompress = Checker.isNeedCompress(mLeastCompressSize, path)
                    if (isNeedCompress) {
                        val compressUtils = CompressUtils()
                        val bitmap = compressUtils.decodeFile(path)
                        /* file = getImageCacheFile(context, Checker.checkSuffix(path));*/compressUtils
                                .compressImage(bitmap, mQuality, path)
                    }
                    if (!result.contains(path)) {
                        //添加到已经缓存的集合中
                        result.add(path)
                    }
                    //判断压缩是否已经结束
                    if (mPaths!!.size == result.size) {
                        mHandler.sendMessage(mHandler.obtainMessage(COMPRESS_SUCCESS, result))
                    }
                }
            } else {
                mCompressListener!!.onError(IllegalArgumentException("This path is not a picture and cannot be read..."))
                break
            }
        }
    }

    override fun handleMessage(msg: Message): Boolean {
        if (mCompressListener == null) {
            return false
        }
        when (msg.what) {
            COMPRESS_START -> mCompressListener.onStart()
            COMPRESS_ERROR -> mCompressListener.onError(msg.obj as Throwable)
            COMPRESS_SUCCESS -> mCompressListener.onSuccess(msg.obj as List<String?>)
            else -> {
            }
        }
        return false
    }

    class Builder(private val mContext: Context) {
        val mPaths: MutableList<String>
        var mLeastCompressSize = 100
        var mCompressListener: OnCompressListener? = null

        //输出文件目录
        var mOutFileDir: String? = null

        //默认100
        var mQuality = 100
        fun load(path: String): Builder {
            mPaths.add(path)
            return this
        }

        fun setQuality(quality: Int): Builder {
            mQuality = quality
            return this
        }

        fun load(paths: List<String>): Builder {
            mPaths.addAll(paths)
            return this
        }

        fun setCompressListener(listener: OnCompressListener?): Builder {
            mCompressListener = listener
            return this
        }

        fun setOutFileDir(outFileDir: String?): Builder {
            mOutFileDir = outFileDir
            return this
        }

        //小于100，忽略图片压缩
        fun ignoreCompress(size: Int): Builder {
            mLeastCompressSize = size
            return this
        }

        private fun build(): ImageCompressUtils {
            return ImageCompressUtils(this)
        }

        fun startCompress() {
            build().startCompress()
        }

        init {
            mPaths = ArrayList()
        }
    }

    companion object {
        //开始
        private const val COMPRESS_START = 101

        //成功
        private const val COMPRESS_SUCCESS = 102

        //失败
        private const val COMPRESS_ERROR = 103
        private const val DEFAULT_DISK_CACHE_DIR = "basic_disk_cache"

        @JvmStatic
        fun with(context: Context): Builder {
            return Builder(context)
        }
    }

    init {
        mPaths = builder.mPaths
        mLeastCompressSize = builder.mLeastCompressSize
        mOutFileDir = builder.mOutFileDir
        mCompressListener = builder.mCompressListener
        mQuality = builder.mQuality
        mHandler = Handler(Looper.getMainLooper(), this)
    }
}